File: src/horse_algos/graph.py
from copy import deepcopy


class Graph:
    """Graph class with DSU and node-cut functionality"""

    def __init__(
        self,
        adjMatrix: list[list[int]] = None, # type: ignore
        nodeValues: list[int] = None, # type: ignore
        adjList: list[list[int]] = None, # type: ignore
        infSet: set[int] = None, # type: ignore
    ):
        """Initializes the Graph class with either an adjacency matrix or an adjacency list.

        Args:
            adjMatrix: The adjacency matrix of the graph.
            nodeValues: The values associated with each node.
            adjList: The adjacency list of the graph.
            infSet: The set of nodes that cannot be removed from the graph.

        Raises:
            ValueError: If nodeValues is not provided, or if neither adjMatrix nor adjList is provided.
        """
        if nodeValues is None:
            raise ValueError("nodeValues must be provided.")

        self.nodeValues = nodeValues
        self.infSet = infSet if infSet is not None else set()

        if adjMatrix is not None:
            self.adjMatrix = adjMatrix
            self.adjList = (
                adjList if adjList is not None else self._adjMatrixToAdjList(adjMatrix)
            )
        elif adjList is not None:
            self.adjList = adjList
            self.adjMatrix = self._adjListToAdjMatrix(adjList)
        else:
            raise ValueError("Either adjMatrix or adjList must be provided.")

        n = len(self.nodeValues)

        # parent[i] = i means it's the root of its own set
        self.parent = list(range(n))
        # rank helps keep the tree balanced
        self.rank = [0] * n
        self.is_active = [True] * n
        # history stores (parent_node, child_node, rank_was_increased)
        self.history = []

    def _adjMatrixToAdjList(self, adjMatrix: list[list[int]]) -> list[list[int]]:
        """Converts an adjacency matrix to an adjacency list.

        Args:
            adjMatrix: The adjacency matrix to convert.

        Returns:
            The corresponding adjacency list.
        """
        adjList = [[] for _ in range(len(adjMatrix))]
        for i in range(len(adjMatrix)):
            for j in range(len(adjMatrix[i])):
                if adjMatrix[i][j] == 1:
                    adjList[i].append(j)
        return adjList

    def _adjListToAdjMatrix(self, adjList: list[list[int]]) -> list[list[int]]:
        """Converts an adjacency list to an adjacency matrix.

        Args:
            adjList: The adjacency list to convert.

        Returns:
            The corresponding adjacency matrix.
        """
        n = len(adjList)
        adjMatrix = [[0] * n for _ in range(n)]
        for i in range(n):
            for neighbor in adjList[i]:
                adjMatrix[i][neighbor] = 1
        return adjMatrix

    def value(self, vertex: int) -> int:
        """Returns the value of the given vertex."""
        return self.nodeValues[vertex]

    def includedValue(self, s: int, t: int, cutset: set[int]) -> int:
        """Returns the total value of the component that includes vertex
        s, given that the vertices in cutset are removed from the graph.

        Args:
            s: The vertex whose component value we want to calculate.
            t: The other vertex (not used in this function, but included for consistency with path).
            cutset: A set of vertices that are removed from the graph.
        """
        visited = set()
        stack = [s]
        totalValue = 0
        while stack:
            vertex = stack.pop()
            if vertex in visited or vertex in cutset:
                continue
            visited.add(vertex)
            totalValue += self.value(vertex)
            for neighbor in self.adjList[vertex]:
                stack.append(neighbor)
        return totalValue

    def find(self, x):
        """Finds the root of the set containing x without path compression."""
        while x != self.parent[x]:
            x = self.parent[x]
        return x

    def unite(self, a, b):
        """Unites sets containing a and b using Union by Rank."""
        root_a = self.find(a)
        root_b = self.find(b)

        if root_a != root_b:
            # We want root_a to be the one with the higher rank
            if self.rank[root_a] < self.rank[root_b]:
                root_a, root_b = root_b, root_a

            rank_increased = False
            if self.rank[root_a] == self.rank[root_b]:
                self.rank[root_a] += 1
                rank_increased = True

            # Make root_a the parent of root_b
            self.parent[root_b] = root_a
            self.history.append(
                {"u": root_a, "v": root_b, "rank_increased": rank_increased}
            )
            return True
        return False
      
    def uniteBySeparator(self, s: int, separator: set[int]):
        """Unites all nodes on the source side of the separator with s."""
        visited = set()
        stack = [s]
        while stack:
            vertex = stack.pop()
            if vertex in visited or vertex in separator:
                continue
            visited.add(vertex)
            self.unite(s, vertex)
            for neighbor in self.adjList[vertex]:
                stack.append(neighbor)
      
    def deactivate(self, v):
        """Deactivates a vertex, effectively removing it from the graph."""
        self.is_active[v] = False
        # Store a clear type flag to avoid KeyErrors during undo
        self.history.append({"type": "deactivate", "node": v})

    def undo(self, target_size):
        """Rolls back the DSU and activation state to a specific history size."""
        while len(self.history) > target_size:
            m = self.history.pop()
            
            if m.get("type") == "deactivate":
                self.is_active[m["node"]] = True
            else:
                # This is a DSU union entry
                self.parent[m["v"]] = m["v"]
                if m["rank_increased"]:
                    self.rank[m["u"]] -= 1

    def generateFlowMatrix(self, s: int, t: int):
        """Generates a flow matrix for the graph where each node has a capacity of 1,
        unless it is in the same DSU component as s, t, or a node in infSet,
        in which case it has infinite capacity.
        Edges between nodes have infinite capacity.
        DSU unions are respected by adding infinite capacity edges between united nodes.

        Args:
            s: The source node.
            t: The sink node.

        Returns:
            A 2D list representing the flow matrix.
        """
        n = len(self.adjMatrix)
        flowMatrix: list[list[int | float]] = [[0] * (2 * n) for _ in range(2 * n)]
        
        root_s = self.find(s)
        root_t = self.find(t)
        inf_roots = {self.find(node) for node in self.infSet}
        inf_roots.add(root_s)
        inf_roots.add(root_t)

        for i in range(n):
            if not self.is_active[i]:
                continue
            
            # Nodes in components of s, t, or infSet have infinite capacity (cannot be cut)
            flowMatrix[2 * i][2 * i + 1] = float("inf") if self.find(i) in inf_roots else 1
            
            # Respect DSU unions by connecting nodes to their root with infinite capacity
            root_i = self.find(i)
            if root_i != i:
                if self.is_active[root_i]:
                    flowMatrix[2 * i + 1][2 * root_i] = float("inf")
                    flowMatrix[2 * root_i + 1][2 * i] = float("inf")

            for j in range(n):
                # Only add edges between active nodes
                if self.adjMatrix[i][j] == 1 and self.is_active[j]:
                    flowMatrix[2 * i + 1][2 * j] = float("inf")
                    
        return flowMatrix


def path(graph: Graph, a: int, b: int, cutset: set[int]) -> bool:
    """Returns true if there exists a path in the graph between a and b
    that does not include any of the vertices in cutset or inactive nodes.
    False otherwise.

    Args:
        graph: The graph to check for a path.
        a: The starting vertex.
        b: The ending vertex.
        cutset: A set of vertices that cannot be included in the path.
    """

    adjList = graph.adjList
    isActive = graph.is_active
    visited = set()
    stack = [a]

    while stack:
        vertex = stack.pop()
        if vertex == b:
            return True
        if vertex in visited or vertex in cutset or not isActive[vertex]:
            continue
        visited.add(vertex)
        for neighbor in adjList[vertex]:
            stack.append(neighbor)

    return False


def minSeparator(graph: Graph, s: int, t: int) -> set[int]:
    """Returns a minimum separator between s and t in the graph."""  
    flowMatrix = graph.generateFlowMatrix(s, t)
    num_nodes = len(flowMatrix)

    def bfs(source, sink, parent):
        visited = [False] * num_nodes
        queue = [source]
        visited[source] = True
        while queue:
            u = queue.pop(0)
            for v in range(num_nodes):
                if not visited[v] and flowMatrix[u][v] > 0:
                    queue.append(v)
                    visited[v] = True
                    parent[v] = u
                    if v == sink:
                        return True
        return False

    # 1. Source is s_out (2*s + 1), Sink is t_in (2*t)
    source_node = 2 * s + 1
    sink_node = 2 * t

    # 2. Edmonds-Karp to saturate the network
    parent = [-1] * num_nodes
    while bfs(source_node, sink_node, parent):
        path_flow = float("inf")
        curr = sink_node
        while curr != source_node:
            path_flow = min(path_flow, flowMatrix[parent[curr]][curr])
            curr = parent[curr]

        v = sink_node
        while v != source_node:
            u = parent[v]
            flowMatrix[u][v] -= path_flow
            flowMatrix[v][u] += path_flow
            v = parent[v]
        parent = [-1] * num_nodes

    # 3. Find the farthest cut (Important Separator):
    # Find all nodes that can reach the sink in the residual graph
    can_reach_t = [False] * num_nodes
    queue = [sink_node]
    can_reach_t[sink_node] = True

    # Backward BFS on the residual graph
    while queue:
        v = queue.pop(0)
        for u in range(num_nodes):
            # If there is capacity from u to v, then v can be reached from u
            if not can_reach_t[u] and flowMatrix[u][v] > 0:
                can_reach_t[u] = True
                queue.append(u)

    # 4. The separator S is the set of nodes where
    # the 'in' node cannot reach t, but the 'out' node CAN reach t.
    min_cut = set()
    for i in range(len(graph.adjMatrix)):
        if i == s or i == t:
            continue
        # If the internal edge (in -> out) is saturated and is the bottleneck
        if not can_reach_t[2 * i] and can_reach_t[2 * i + 1]:
            min_cut.add(i)

    return min_cut

File: cpp/graph.cpp
#include "graph.hpp"
#include <algorithm>
#include <queue>
#include <limits>
#include <iostream>

Graph::Graph(const std::vector<std::vector<int>>& adjList, const std::vector<int>& nodeValues, const std::set<int>& infSet)
    : adjList(adjList), nodeValues(nodeValues), infSet(infSet) {
    int n = nodeValues.size();
    parent.resize(n);
    rank.resize(n, 0);
    is_active.resize(n, true);
    for (int i = 0; i < n; ++i) parent[i] = i;
}

int Graph::find(int x) {
    while (x != parent[x]) {
        x = parent[x];
    }
    return x;
}

bool Graph::unite(int a, int b) {
    int root_a = find(a);
    int root_b = find(b);
    if (root_a != root_b) {
        if (rank[root_a] < rank[root_b]) std::swap(root_a, root_b);
        bool rank_increased = false;
        if (rank[root_a] == rank[root_b]) {
            rank[root_a]++;
            rank_increased = true;
        }
        parent[root_b] = root_a;
        history.push_back({"union", root_a, root_b, rank_increased, -1});
        return true;
    }
    return false;
}

void Graph::deactivate(int v) {
    is_active[v] = false;
    history.push_back({"deactivate", -1, -1, false, v});
}

void Graph::undo(int target_size) {
    while (history.size() > (size_t)target_size) {
        HistoryEntry m = history.back();
        history.pop_back();
        if (m.type == "deactivate") {
            is_active[m.node] = true;
        } else {
            parent[m.v] = m.v;
            if (m.rank_increased) {
                rank[m.u]--;
            }
        }
    }
}

int Graph::includedValue(int s) {
    std::set<int> visited;
    std::vector<int> stack = {s};
    int totalValue = 0;
    while (!stack.empty()) {
        int vertex = stack.back();
        stack.pop_back();
        if (visited.count(vertex) || !is_active[vertex]) continue;
        visited.insert(vertex);
        totalValue += nodeValues[vertex];
        for (int neighbor : adjList[vertex]) {
            stack.push_back(neighbor);
        }
    }
    return totalValue;
}

void Graph::uniteBySeparator(int s, const std::set<int>& separator) {
    std::set<int> visited;
    std::vector<int> stack = {s};
    while (!stack.empty()) {
        int vertex = stack.back();
        stack.pop_back();
        if (visited.count(vertex) || separator.count(vertex) || !is_active[vertex]) continue;
        visited.insert(vertex);
        unite(s, vertex);
        for (int neighbor : adjList[vertex]) {
            stack.push_back(neighbor);
        }
    }
}

bool Graph::hasPath(int a, int b, const std::set<int>& cutset) {
    std::set<int> visited;
    std::vector<int> stack = {a};
    while (!stack.empty()) {
        int vertex = stack.back();
        stack.pop_back();
        if (visited.count(vertex) || cutset.count(vertex) || !is_active[vertex]) continue;
        if (vertex == b) return true;
        visited.insert(vertex);
        for (int neighbor : adjList[vertex]) {
            stack.push_back(neighbor);
        }
    }
    return false;
}

std::vector<std::vector<double>> Graph::generateFlowMatrix(int s, int t) {
    int n = nodeValues.size();
    std::vector<std::vector<double>> flowMatrix(2 * n, std::vector<double>(2 * n, 0));
    
    int root_s = find(s);
    int root_t = find(t);
    std::set<int> inf_roots;
    for (int node : infSet) inf_roots.insert(find(node));
    inf_roots.insert(root_s);
    inf_roots.insert(root_t);

    double inf = std::numeric_limits<double>::infinity();

    for (int i = 0; i < n; ++i) {
        if (!is_active[i]) continue;
        
        flowMatrix[2 * i][2 * i + 1] = inf_roots.count(find(i)) ? inf : 1.0;
        
        int root_i = find(i);
        if (root_i != i && is_active[root_i]) {
            flowMatrix[2 * i + 1][2 * root_i] = inf;
            flowMatrix[2 * root_i + 1][2 * i] = inf;
        }

        for (int neighbor : adjList[i]) {
            if (is_active[neighbor]) {
                flowMatrix[2 * i + 1][2 * neighbor] = inf;
            }
        }
    }
    return flowMatrix;
}

std::set<int> Graph::minSeparator(int s, int t) {
    auto flowMatrix = generateFlowMatrix(s, t);
    int num_nodes = flowMatrix.size();
    int source_node = 2 * s + 1;
    int sink_node = 2 * t;
    double inf = std::numeric_limits<double>::infinity();

    auto bfs = [&](std::vector<int>& parent) {
        std::fill(parent.begin(), parent.end(), -1);
        std::queue<int> q;
        q.push(source_node);
        parent[source_node] = source_node;
        while (!q.empty()) {
            int u = q.front(); q.pop();
            for (int v = 0; v < num_nodes; ++v) {
                if (parent[v] == -1 && flowMatrix[u][v] > 0) {
                    parent[v] = u;
                    if (v == sink_node) return true;
                    q.push(v);
                }
            }
        }
        return false;
    };

    std::vector<int> parent(num_nodes);
    while (bfs(parent)) {
        double path_flow = inf;
        for (int v = sink_node; v != source_node; v = parent[v]) {
            path_flow = std::min(path_flow, flowMatrix[parent[v]][v]);
        }
        if (path_flow == inf) break;
        for (int v = sink_node; v != source_node; v = parent[v]) {
            int u = parent[v];
            if (flowMatrix[u][v] != inf) flowMatrix[u][v] -= path_flow;
            if (flowMatrix[v][u] != inf) flowMatrix[v][u] += path_flow;
        }
    }

    std::vector<bool> can_reach_t(num_nodes, false);
    std::queue<int> q;
    q.push(sink_node);
    can_reach_t[sink_node] = true;
    while (!q.empty()) {
        int v = q.front(); q.pop();
        for (int u = 0; u < num_nodes; ++u) {
            if (!can_reach_t[u] && flowMatrix[u][v] > 0) {
                can_reach_t[u] = true;
                q.push(u);
            }
        }
    }

    std::set<int> min_cut;
    for (int i = 0; i < (int)adjList.size(); ++i) {
        if (i == s || i == t) continue;
        if (!can_reach_t[2 * i] && can_reach_t[2 * i + 1]) {
            min_cut.insert(i);
        }
    }
    return min_cut;
}
